server/src/main/java/org/elasticsearch/index/similarity/SimilarityService.java

```diff
@@ -171,4 +194,231 @@ public Similarity get(String name) {
             return (fieldType != null && fieldType.similarity() != null) ? fieldType.similarity().get() : defaultSimilarity;
         }
     }
+
+    static void validateSimilarity(Version indexCreatedVersion, Similarity similarity) {
+        try {
+            validateScoresArePositive(indexCreatedVersion, similarity);
+            validateScoresDoNotDecreaseWithFreq(indexCreatedVersion, similarity);
+            validateScoresDoNotIncreaseWithNorm(indexCreatedVersion, similarity);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static class SingleNormLeafReader extends LeafReader {
+
```

Comment:
I do wonder if this class should be in a separate utility file somewhere?  It's a third of the entire file length.  On the other hand, this is all going away in master, so maybe it doesn't matter.

Reply:
Good point. I'll leave things this way though as master doesn't have the issue so it's not a big deal like you said.